### DISCUSSION TITLE
x-model error when removing object from array

### DISCUSSION THREAD

**Discussion by alex-bntz** [alpinejs/alpine#3088] [7. Bugs (archived)] [answered]
Hi there,

There is a weird behavior going on with x-model inside a loop when deleting an object form the array.

So the idea is this:
```
[
    {
        x-model 1
        x-model 2
    },
    {
        x-model 1
        x-model 2
    }
]
```

and in the loop I connect the models like this:
```
dataArray[loopIndex]['x-model 1']
dataArray[loopIndex]['x-model 2']
```

and that works just fine.. The issue now is, whenever I remove a **specific** object from the array, I get errors in the console
```
Cannot read properties of undefined (reading 'x-model 1')
Cannot read properties of undefined (reading 'x-model 2')
```

I've used 2 different methods to remove an object from the array:
`dataArray.splice(loopIndex, 1)`
or
```
dataArray = dataArray.filter((object, objectIndex) => {
    return loopIndex !== objectIndex;
});
```
Both those ways work but either way I get errors in the console for each x-model in the object.

I've created a codepen with minimal reproduction: https://codepen.io/alexbtnz/pen/QWmrrMr

Any help or workaround much appreciated. :)

**Comment by SimoTod** [comment, marked as answer]
The issue is that you use index as a key instead of a real unique id from the data. Index change at every rendering so is not even suitable to track items when the state is important.

**Reply by SimoTod** [reply]
For example: https://codepen.io/SimoTod/pen/vYRjagM

**Reply by alex-bntz** [reply]
That's actually it @SimoTod.
I added a random uuid as id of the object and followed your approach and I can confirm it worked like a charm!
Thank you very much! I could find that issue anywhere online so it's still nice to be here for anyone else who does the same mistake 😄

**Reply by alex-bntz** [reply]
I just noticed actually, it still throws those errors if you try to remove the last item. Any thoughts on that?

**Reply by SimoTod** [reply]
I'm not sure why it does it because technically your code should work however it's a bit unusual. obj is already data['index'] so you usually prefer the former and have stuff like obj['firstName']

**Reply by alex-bntz** [reply]
Changing `dataArray[index].firstName` to `obj.firstName` indeed fixes that remaining issue. I think I saw you could connect an x-model to a nested obj only with the first approach but looks like this isn't the case (anymore).

Thanks again @SimoTod 🍻
